calculation/sha_calc/sha_calc/directivity/bea20/directivity.py

```diff
@@ -31,45 +32,363 @@ def compute_directivity_effects(
+            lon_lat_depth, planes, nominal_strike, hypocentre_origin=True
+        )
+        rx_end2, ry_end2 = src_site_dist.calc_rx_ry_GC2(
+            lon_lat_depth, planes, nominal_strike2, hypocentre_origin=True
+        )
+        s_max = (min(ry_end, ry_end2)[0], max(ry_end, ry_end2)[0])
+
+        # Gets the plane index of the hypocentre
+        plane_index = planes_index[index]
+
+        # Trig to calculate extra features of the fault for directivity based on plane info
+        z_tor = planes[plane_index]["dtop"]
+        dip = planes[plane_index]["dip"]
+        d_bot = z_tor + planes[plane_index]["width"] * math.sin(math.radians(dip))
+        t_bot = z_tor / math.tan(math.radians(dip)) + planes[plane_index][
+            "width"
+        ] * math.cos(math.radians(dip))
+        d = (planes[plane_index]["dhyp"] - z_tor) / math.sin(math.radians(dip))
+
+        # Use the bea20 model to work out directivity (fd) at the given sites
+        fd, fdi, phi_red, phi_redi, predictor_functions, other = bea20.bea20(
+            mag, ry, rx, s_max, d, t_bot, d_bot, rake, dip, period
+        )
+
+        if fdi_average.size == 0:
+            (
+                fd_average,
+                fdi_average,
+                phi_red_average,
+                phi_redi_average,
+                predictor_functions_average,
+                other_average,
+            ) = (
+                fd,
+                fdi,
+                phi_red,
+                phi_redi,
+                predictor_functions,
+                other,
+            )
+        else:
+            (fd_average, fdi_average, phi_red_average, phi_redi_average,) = (
+                np.add(fd_average, fd),
+                np.add(fdi_average, fdi),
+                np.add(phi_red_average, phi_red),
+                np.add(phi_redi_average, phi_redi),
+            )
+            for key, value in predictor_functions.items():
+                predictor_functions_average[key] = np.add(
+                    predictor_functions_average[key], predictor_functions[key]
+                )
+            for key, value in other.items():
+                other_average[key] = np.add(other_average[key], other[key])
+
+    (fd_average, fdi_average, phi_red_average, phi_redi_average,) = (
+        np.divide(fd_average, n_hypo),
+        np.divide(fdi_average, n_hypo),
+        np.divide(phi_red_average, n_hypo),
+        np.divide(phi_redi_average, n_hypo),
+    )
+    for key, value in predictor_functions_average.items():
+        predictor_functions_average[key] = np.divide(
+            predictor_functions_average[key], n_hypo
+        )
+    for key, value in other_average.items():
+        other_average[key] = np.divide(other_average[key], n_hypo)
+
+    return (
+        fd_average,
+        fdi_average,
+        phi_red_average,
+        phi_redi_average,
+        predictor_functions_average,
+        other_average,
+    )
+
+
+def directivity_plots(
```

Comment:
I would suggest making the validation plotting functionality scripts instead of functions, 
- 1 script for generating the non-hypocentre validation plots (i.e. the one with 6 sub-figures)
- 1 script for generating the hypocentre-averaring plots (both average & individiuals, but only fdi (unless there is a good reason to plot the others?))
- Put shared plotting functionality into a plots.py module

Reply:
And put said scripts along with the output plots into a `validation` folder